FMA for contract version upgrade (#69)
ethereum-optimism/design-docs: security/fma-contract-version-upgrade.md

<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Introduction](#introduction)
- [Failure Modes and Recovery Paths](#failure-modes-and-recovery-paths)
  - [Missing side effect on `.selector` access bug (Bug fixed after 0.8.15)](#missing-side-effect-on-selector-access-bug-bug-fixed-after-0815)
  - [Storage write removal before conditional termination (Bug fixed after 0.8.15)](#storage-write-removal-before-conditional-termination-bug-fixed-after-0815)
  - [Change in Default EVM version between different Solidity versions](#change-in-default-evm-version-between-different-solidity-versions)
- [Audit Requirements](#audit-requirements)
- [Action Items](#action-items)
- [Appendix](#appendix)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

|                     |                                    |
| ------------------- | ---------------------------------- |
| Author              | Michael Amadi                      |
| Created at          | 2024-08-20                         |
| Needs Approval From | _Security Reviewer Name_           |

> | Other Reviewers     | _Reviewer Name 1, Reviewer Name 2_ |

Reviewer: You can put needs approval from as Matt Solomon and put me (Mark Tyneway) as other reviewer

| Status              | Draft         |

> [!NOTE]
> 📢 Remember:
>
> - The single approver in the “Need Approval From” must be from the Security team.
> - Maintain the “Status” property accordingly.

> [!TIP]
> Guidelines for writing a good analysis, and what the reviewer will look for:
>
> - Show your work: Include steps and tools for each conclusion.
> - Completeness of risks considered.
> - Include both implementation and operational failure modes
> - Provide references to support the reviewer.
> - The size of the document will likely be proportional to the project's complexity.
> - The ultimate goal of this document is to identify action items to improve the security of the project. The FMA review process can be accelerated by proactively identifying action items during the writing process.

## Introduction

> This document covers solidity version upgrade of all OP Stack smart contracts from 0.8.15 code to 0.8.25.

Reviewer 2: Per our discord conversation: We might want some of the tstore features in 0.8.27, so I would suggest not assumeing 0.8.25 is the target, and adding an FMA item like "New solidity version is too recent and contains unknown bugs" that discusses this tradeoff

Below are references for this project:

- [Issue 11527: Smart Contracts: Update all 0.8.15 code to 0.8.25](https://github.com/ethereum-optimism/optimism/issues/11527).

> ## Failure Modes and Recovery Paths

Reviewer 2: Can you add some intro text to this section explaining why the below solidity bug fixes are present? In other words, there have likely been a lot of bug fixes since 0.8.15, and we should definitely not enumerate all of them here, so the text would explain why this subset is the one listed.

I'd also suggest a slight tweak to the format:

```markdown
## Failure Modes and Recovery Paths

### Solidity Bugs Fixed

This section lists solidity bugs that were fixed as failure modes and explains why these fixes might be a risk for us. This set of bug fixes was chosen because....

#### Missing side effect on `.selector` access bug (Bug fixed after 0.8.15)

etc...

### Other Failure Modes

#### New bugs introduced in later solidity versions

#### Stack too deep

etc....
```

Reviewer 2: The above comment also hints at another suggestion: we should have a failure mode for exacerbating our [stack too deep](https://github.com/ethereum-optimism/optimism/issues/11093) problem. Currently our codebase only compile with the legacy pipeline with optimizer on. We cannot compile with optimizer off, nor with via-ir, as both result in stack too deep. We need to make sure whatever version we choose does not also give us stack too deep for our current build settings

### Missing side effect on `.selector` access bug (Bug fixed after 0.8.15)

- **Description:**
  - Introduced: v0.6.2
  - Fixed: v0.8.21
  - Explanation: The legacy compiler assumes that conditionals used to determine which selector is used/removed have no side effects. So if the result of the conditional is known at compile time the legacy compiler will hardcode the corresponding selector (since selectors are known at compile time). This bug does no exist when the conditional is used with other types of constants.
- **Risk Assessment:**
  - Severity: High (if present)
  - Likelihood: Low (and mostly present in anti-patterns like using conditionals that have side effects)
- **Mitigations:** Using a solidity version above 0.8.20, if lower versions must be used, conditionals should have no side effects.
- **Detection:** Confidence on the absence of this bug can be stronger with robust unit tests that check that all side effects of a call happens and correctly.

> - **Recovery Path(s)**: Redeployment and upgrade of the affected contracts

Reviewer: I don't believe we have this issue in our codebase

```
$ git grep -rin '.selector' src 

src/L1/DelayedVetoable.sol:10:///         Additionally, this contract cannot be used to forward calls with data beginning with the function selector
src/L1/L1ERC721Bridge.sol:106:        bytes memory message = abi.encodeWithSelector(
src/L1/L1ERC721Bridge.sol:107:            L2ERC721Bridge.finalizeBridgeERC721.selector, _remoteToken, _localToken, _from, _to, _tokenId, _extraData
src/L2/L1Block.sol:141:                mstore(0x00, 0x3cc50b45) // 0x3cc50b45 is the 4-byte selector of "NotDepositor()"
src/L2/L1Block.sol:142:                revert(0x1C, 0x04) // returns the stored 4-byte selector from above
src/L2/L2ERC721Bridge.sol:114:        bytes memory message = abi.encodeWithSelector(
src/L2/L2ERC721Bridge.sol:115:            L1ERC721Bridge.finalizeBridgeERC721.selector, remoteToken, _localToken, _from, _to, _tokenId, _extraData
src/cannon/PreimageOracle.sol:272:                // Store the "ShaFailed()" error selector.
src/cannon/PreimageOracle.sol:303:                // Store the "InvalidProof()" error selector.
src/cannon/PreimageOracle.sol:518:                // Store "InvalidInputSize()" error selector
src/dispute/FaultDisputeGame.sol:213:        // - 0x04 selector
src/dispute/FaultDisputeGame.sol:221:                // Store the selector for `BadExtraData()` & revert
src/legacy/L1ChugSplashProxy.sol:35:            owner.staticcall(abi.encodeWithSelector(IL1ChugSplashDeployer.isUpgrading.selector));
src/legacy/LegacyMintableERC20.sol:48:        bytes4 secondSupportedInterface = ILegacyMintableERC20.l1Token.selector ^ ILegacyMintableERC20.mint.selector
src/legacy/LegacyMintableERC20.sol:49:            ^ ILegacyMintableERC20.burn.selector;
src/libraries/SafeCall.sol:123:                // Store the "Error(string)" selector in scratch space.
src/universal/CrossDomainMessenger.sol:189:            _data: abi.encodeWithSelector(
src/universal/CrossDomainMessenger.sol:190:                this.relayMessage.selector, messageNonce(), msg.sender, _target, msg.value, _minGasLimit, _message
src/universal/StandardBridge.sol:333:            _message: abi.encodeWithSelector(this.finalizeBridgeETH.selector, _from, _to, _amount, _extraData),
src/universal/StandardBridge.sol:378:            _message: abi.encodeWithSelector(
src/universal/StandardBridge.sol:379:                this.finalizeBridgeERC20.selector,

```

### Storage write removal before conditional termination (Bug fixed after 0.8.15)

- **Description:**

  - Introduced: v0.8.13
  - Fixed: v0.8.17
  - Explanation: Incorrect assumption that a storage write is redundant so the optimizer removes it. But this assumption can be wrong when the developer intended otherwise. A storage write is considered redundant when the storage slot is written to twice within an execution context without that slot being read before the second write or execution unconditionally terminates before the slot is read again.

> 

Reviewer: This is a bug fix. We may have been exposed to the bug but unlikely. The way that this could hurt us is if we depended on the buggy behavior

Reviewer 2: I vaguely recall talking with @maurelian about this a while ago, and think he said we checked our codebase when this was announced? However it's been 2 years since this bug was fixed so we may have introduced this bug at some point after. 

Regardless, agree with @tynes that depending on the buggy behavior is the primary risk, but that is unlikely

Blog post for reference: https://soliditylang.org/blog/2022/09/08/solidity-0.8.17-release-announcement/

Reviewer 3: No, I don't recall doing that check. 

The one that I recall we did a review for was nested [Size Check Bug in Nested Calldata Array ABI-Reencoding](https://soliditylang.org/blog/2022/05/17/calldata-reencode-size-check-bug/).

Reviewer 2: Got it, so we probably should add an action to track verifying this

    According to the solidity blog, this bug can only occur if a contract contains pattern:

    1. A storage write. Note that the write may still be removed even if it happens only conditionally or within a call to a function that ends up being inlined.
    2. A call to a function that conditionally terminates using inline assembly as described above, but also has a different code path that returns to the caller.
    3. Any continuing control flow path does one of the following:
       - It overwrites the storage write in (1).
       - It reverts.

- **Risk Assessment:**
  - Severity: High (if present)
  - Likelihood: Low
- **Mitigations:** Use a solidity version above 0.8.17.
- **Detection:** If a solidity contract (using solidity version earlier than 0.8.17) follows the pattern described above, its potentially vulnerable to this.
- **Recovery Path(s)**: Redeployment and upgrade of the affected contracts

### Change in Default EVM version between different Solidity versions

> - **Description:** Version 0.8.25 sets the default evm version to cancun

Reviewer: This is no problem

- **Risk Assessment:**
  - Severity: Low/Medium/High
  - Likelihood: Medium
- **Mitigations:** As at 0.8.15, the default EVM version is 'London', if no explicit version override is declared, 'Cancun' will be used rather which if not intended can lead to code that either does not deploy or reverts un-ideally at runtime even though non-fork tests might have passed. A mitigation is explicitly declaring the intended EVM version to use for compilation regardless of if its the same as the default the compiler uses. That way, a new compiler version that changes the default EVM version does not change the one that is intended to be used. In this scenario however, this is not possible because the intended EVM version is the same as the compiler's EVM version default.
- **Detection:** Running unit tests as fork test also.
- **Recovery Path(s)**: Redeployment and upgrade of the affected contracts

## Audit Requirements

> _Will this project require an audit according to the guidance in [OP Labs Audit Framework: When to get external security review and how to prepare for it](https://gov.optimism.io/t/op-labs-audit-framework-when-to-get-external-security-review-and-how-to-prepare-for-it/6864)? Please explain your reasoning._

Reviewer 2: Let's fill this section out. My gut is that this does not need an audit, though offhand I'm not sure how it fits into the audit framework

Reviewer 2: I see the thumbs up but looks like this rationale is still missing?

## Action Items

Below is what needs to be done before launch to reduce the chances of the above failure modes occurring, and to ensure they can be detected and recovered from:

> - [ ] Resolve all comments on this document and incorporate them into the document itself (Assignee: document author)

Reviewer 2: Any other action items we add here? Two that come to mind are testing the stack too deep and code size situations from whatever new version we choose (ref https://github.com/ethereum-optimism/design-docs/pull/69/files?diff=unified&w=0#r1746167621)

Reviewer 2: New actions LGTM, one more action could be to decide on a solidity version, based on the new FMA item

> # [Project Name]: Failure Modes and Recovery Path Analysis

Reviewer 2: ```suggestion
# Solidity Version Upgrade: Failure Modes and Recovery Path Analysis
```

    - [Storage write removal before conditional termination (Bug fixed after 0.8.15)](#storage-write-removal-before-conditional-termination-bug-fixed-after-0815)
  - [Other Failure Modes](#other-failure-modes)
    - [Change in Default EVM version between different Solidity versions](#change-in-default-evm-version-between-different-solidity-versions)
    - [Chosen solidity version is too recent and might contain unknown bugs](#chosen-solidity-version-is-too-recent-and-might-contain-unknown-bugs)
    - [Stack too deep errors during compilation](#stack-too-deep-errors-during-compilation)
    - [Contracts exceeding code size when compiled](#contracts-exceeding-code-size-when-compiled)
- [Audit Requirements](#audit-requirements)
- [Action Items](#action-items)
- [Appendix](#appendix)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

|                     |               |
| ------------------- | ------------- |
| Author              | Michael Amadi |
| Created at          | 2024-08-20    |
| Needs Approval From | Matt Solomon  |
| Other Reviewers     | Mark Tyneway  |

> | Status              | Draft         |

Reviewer 2: ```suggestion
| Status              | In Review     |
```

  - Severity: -
  - Likelihood: -
- **Mitigations:** -
- **Detection:** During compilation
- **Recovery Path(s)**: Code modification (local variable scoping, local variables in memory etc), build settings modification or/and usage of a different compiler version.

#### Contracts exceeding code size when compiled

- **Description:** There are a few OP Stack contracts that are close to the contract code size limit. We have to ensure that the chosen compiler version to upgrade to compiles all OP Stack contracts successfully and without any exceeding the contract code size limit.
- **Risk Assessment:**
  - Severity: -
  - Likelihood: -
- **Mitigations:** -
- **Detection:** During compilation
- **Recovery Path(s)**: Code modification (runtime code size optimization), build settings modification or/and usage of a different compiler version.

## Audit Requirements

_Will this project require an audit according to the guidance in [OP Labs Audit Framework: When to get external security review and how to prepare for it](https://gov.optimism.io/t/op-labs-audit-framework-when-to-get-external-security-review-and-how-to-prepare-for-it/6864)? Please explain your reasoning._

> This project would not require an audit. An audit would require a bytecode audit which is not practical. Also, historically, we trust compiler generated code and don't audit bytecode so we are continuing to do the same here.

Reviewer 2: small nit to clarify
```suggestion
This project would not require an audit. Because we don't expect to change Solidity code aside from the pragmas, an audit would have to review the bytecode diff, which is not practical. Also, historically, we trust compiler generated code and don't audit bytecode so we are continuing to do the same here.
```

## Action Items

Below is what needs to be done before launch to reduce the chances of the above failure modes occurring, and to ensure they can be detected and recovered from:

- [ ] Decide on the solidity version to upgrade to
- [ ] Resolve all comments on this document and incorporate them into the document itself (Assignee: document author)
- [ ] Contracts compile successfully with upgraded solidity version and without stack too deep error(s)

> - [ ] Contracts compile successfully with upgraded solidity version and without any contract intended to be deployed exceeding the contract code size limit (24576 bytes)

Reviewer 2: Let's make sure each action is assigned to someone so we know who owns which action, I'll defer to @AmadiMichael @tynes for this

Reviewer: @smartcontracts can help here

Reviewer 2: Bump on this, once:
1. the suggested action from https://github.com/ethereum-optimism/design-docs/pull/69/files#r1773672065 is added, and
2. all actions have an assignee

I will approve this and we can transition this to implementing actions
